n3fit/src/n3fit/hyper_optimization/rewards.py

```diff
@@ -14,26 +14,125 @@
+    """Loss function for hyperoptimization based on the distance of
+    the fits of all folds to the first fold
+    """
+    if n3pdfs is None:
+        raise ValueError("fit_distance needs n3pdf models to act upon")
+    xgrid = np.concatenate([np.logspace(-6, -1, 20), np.linspace(0.11, 0.9, 30)])
+    plotting_grids = [xplotting_grid(pdf, 1.6, xgrid) for pdf in n3pdfs]
+    distances = distance_grids(n3pdfs, plotting_grids, 0)
+    # The first distance will obviously be 0
+    # TODO: define this more sensibly, for now it is just a template
+    max_distance = 0
+    for distance in distances:
+        max_distance = max(max_distance, distance.grid_values.max())
+    return max_distance
+
+
+def _set_central_value(n3pdf, model):
+    """Given an N3PDF object and a MetaModel, set the PDF_0 layer
+    to be the central value of the n3pdf object"""
+    from n3fit.backends import operations as op
+
+    # Get the input x
+    for key, grid in model.x_in.items():
+        if key != "integration_grid":
+            input_x = grid.numpy()
+            break
+
+    # Compute the central value of the PDF
+    full_pdf = n3pdf(input_x)
+    cv_pdf = op.numpy_to_tensor(np.mean(full_pdf, axis=0, keepdims=True))
+
+    def central_value(x, training=None):
+        return cv_pdf
+
+    model.get_layer("PDF_0").call = central_value
+    # This model won't be trainable ever again
+    model.trainable = False
+    model.compile()
+
+
+def fit_future_tests(fold_losses, n3pdfs=None, experimental_models=None, **kwargs):
+    """Use the future tests as a metric for hyperopt
+
+    NOTE: this function should only be called once at the end of
+    every hyperopt iteration, as it is destructive for the models
+    """
+    if n3pdfs is None:
+        raise ValueError("fit_future_test needs n3pdf models to act upon")
+    if experimental_models is None:
+        raise ValueError("fit_future_test needs experimental_models to compute chi2")
+
+    from n3fit.backends import MetaModel
+
+    # For the last model (the one containing all data) the PDF covmat doesn't need to be computed
+    # but the mask needs to be flipped in the folding for the appropiate datasets
+    last_model = experimental_models[-1]
+    _set_central_value(n3pdfs[-1], last_model)
+
+    # Loop over all models but the last (our reference!)
+    total_loss = 0.0
+    for n3pdf, exp_model in zip(n3pdfs[:-1], experimental_models[:-1]):
+
+        _set_central_value(n3pdf, exp_model)
+
+        # Get the full input and the total chi2
+        full_input = exp_model.input
+        exp_chi2 = exp_model.compute_losses()["loss"][0]
+
+        # Now update the loss with the PDF covmat
+        for layer in exp_model.get_layer_re(".*_exp$"):
+            # Get the input to the loss layer
+            model_output = MetaModel(full_input, layer.input)
+            # Get the full predictions and generate the PDF covmat
+            y = model_output.predict()[0]  # The first is a dummy-dim
```

Comment:
This is an "old-TF" bug, since (from the discussion in #1405) we are keeping compatibility until the paper is published I will add a workaround so it can be removed once we drop 2.3

(although now I'll have to find out up to which version exactly the workaround is needed...)